frappe/integrations/webhooks.py

```diff
@@ -0,0 +1,27 @@
+# Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
+# MIT License. See license.txt
+
+from __future__ import unicode_literals
+import frappe, requests
+from frappe import _
+
+# Doc Events Webhook
+def doc_event_webhook(doc, method=None, *args, **kwargs):
+	headers = {}
+	filters = {
+		"webhook_doctype": doc.get("doctype"),
+		"webhook_docevent": method
+	}
+	webhooks = frappe.get_all("Webhook", filters=filters)
+	webhook = frappe.get_doc("Webhook", webhooks[0].get("name")) if webhooks and len(webhooks) > 0 else None
+	if webhook:
+		if webhook.webhook_headers:
+			for h in webhook.webhook_headers:
+				if h.get("key") and h.get("value"):
+					headers[h.get("key")] = h.get("value")
+		try:
+			r = requests.post(webhook.request_url, data=doc.as_json(), headers=headers, timeout=5)
```

Comment:
I'm using `doc` here and sending it to the request url as data.

Reply:
I was thinking you can send individual values as parameters. Sometimes you may not want to send the whole doc for security reasons (?) if you want to hide some properties like discount etc.

Reply:
Either ways, if this works for you then fine. Let me merge!

Reply:
Only system managers can add webhooks, so it will be 100% voluntary decision of system manager to communicate to third party apps via webhook. 

Data is expected to be read by webapp and not human directly, so example discount won't actually reach the end user unless third party web app decides.

If user (non system manager) is not allowed to access the `doc`, the wehhook won't be triggered.

It is working as expected, can be merged :)

Reply:
@revant This is a very interesting and very useful to integrate with other applications (mainly if you need "live" updates).

But I also think like @rmehta, you should be able to control the information you send outside erpnext. I think you could add something like "run as user XXX" so you can control the permissions of that user (just an idea for a feature upgrade)

Reply:
I tried creating a new user with access only as blogger (in ERPNext)
This user is not able to access `Sales Invoice` and save/update/submit it. In turn this user won't trigger the Webhook.

It is obeying the frappe permission system as long as it is not customized or `ignore_permission` is set to `True` while triggering `doc_events`

@almeidapaulopt 
> I think you could add something like "run as user XXX" so you can control the permissions of that user (just an idea for a feature upgrade)

As long as system manager doesn't decide to add webhook any data won't be leaked and even if there is a webhook user must have permission to trigger doc_event to trigger webhook.

"run as user" is managed by role and user permission already.

I'll think of something to control data of `doc`.

My case while adding webhook:
As a system manager I need all of the data from doc when events are triggered
As a user I am only permitted to trigger doc_events allowed by role and user permissions.

Reply:
I agree with @almeidapaulopt that we should not pass the full doc. A better design would be to pass specific properties in the child table. (One row per child table). Let me know if you want to do a quick call